BlockFactory: Adding JSON/JavaScript import support

AnmAtAnm opened the pull request:
Updates the BlockFactory main workspace based on the preview block generated from manual construction.

This is a continuation of PR #1216 by @JC-Orozco, rebased on latest `develop` branch. Bring it latest code up to Google styling and standards.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/1235)
<!-- Reviewable:end -->

JC-Orozco commented:
This is a confirmation that I agree to contribute the submitted code authored by me on this push request to the blockly project.

rachel-fenichel commented:
Reviewed 1 of 7 files at r1, 3 of 6 files at r3.
Review status: 4 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[demos/blockfactory/factory.js, line 140 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZivDJKaoga4JmmdPN:-KqZivDKY1eeINyZ8Kpc:b-foo69i) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L140)):*
> ```JavaScript
>   }
> 
>   if (!BlockFactory.updateBlocksFlag) {
> ```

So if the `updateBlocksFlag` is set you just don't update the blocks?

---

*[demos/blockfactory/factory.js, line 143 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZj2Z59J58O3P0ncIr:-KqZj2Z63DSJ0EMhX2af:bazhpt7) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L143)):*
> ```JavaScript
>     var format = document.getElementById('format').value;
>     if (format == 'Manual-JSON') {
>       format = JSON;
> ```

`JSON` should be a string.

---

*[demos/blockfactory/factory.js, line 192 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZie5x1RMDyZh7Yx5D:-KqZie5x1RMDyZh7Yx5E:b-5abjnb) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L192)):*
> ```JavaScript
> 
>   var rawFormat = document.getElementById('format').value;
>   var isManual = rawFormat.substr(0, 6) == 'Manual';
> ```

Where is `isManual` being used?

---

*[demos/blockfactory/factory.js, line 200 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZiSkPX90S1005c1C0:-KqZiSkPX90S1005c1C1:b-1watxd) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L200)):*
> ```JavaScript
>   }
> 
>   switch (rawFormat) {
> ```

Make this a helper function that returns a string or throws an error.

---

*[demos/blockfactory/factory.js, line 232 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfZVoT9-d8ygvtE4f:-KqZfZVoT9-d8ygvtE4g:b-ki6wfx) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L232)):*
>.<

Discussed offline, and we will not have the Manual JavaScript option until we can spin up an interpreter to do the eval.  The code will remain.

---

*[demos/blockfactory/factory.js, line 290 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfSJtTL2V482SlgvC:-KqZfSJtTL2V482SlgvD:b1jfbed) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L290)):*
> ```JavaScript
> 
> /**
>  * Disable link and save buttons if the format is 'Manual', enable otherwise.
> ```

Future thought: if it was parseable from the manual format then it's possible to save and link.

---

*[demos/blockfactory/factory.js, line 330 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfLHi64AO8c3i_v85:-KqZfLHi64AO8c3i_v86:b-ykul0y) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L330)):*
> ```JavaScript
>  */
> BlockFactory.manualEdit = function() {
>   // TODO: Replace these global state flags with parameters passed to the right
> ```

TODO on github and link to issue.

---

*[demos/blockfactory/factory.js, line 335 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfJ_BCjEHdIQubNI-:-KqZfJ_BCjEHdIQubNI0:b-es0i91) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L335)):*
> ```JavaScript
>   BlockFactory.updateBlocksFlagDelayed = true;
>   BlockFactory.updateLanguage();
> }
> ```

Nit: missing newline at end of file.

---

*[demos/blockfactory/factory_utils.js, line 88 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfB8DfJF1nRsYB38u:-KqZfB8DfJF1nRsYB38v:b-utjpwz) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory_utils.js#L88)):*
> ```JavaScript
>     var xml = new BlockDefinitionExtractor().buildBlockFactoryWorkspace(block);
>     Blockly.Xml.domToWorkspace(xml, BlockFactory.mainWorkspace);
>     // Calculate timer to avoid infinite update loops
> ```

Add a TODO and github issue to figure out where the infinite loops are coming from and eventually remove this.

---

*[demos/blockfactory/factory_utils.js, line 91 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfG5sBkaIxhBPZ_Ek:-KqZfG5sBkaIxhBPZ_El:b-mfdj9s) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory_utils.js#L91)):*
> ```JavaScript
>     BlockFactory.updateBlocksFlag = false;
>     setTimeout(
>         function() { BlockFactory.updateBlocksFlagDelayed = false }, 3000);
> ```

nit: semicolon after `false`

---

*[demos/blockfactory/index.html, line 358 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZevZ25MFzGRa584gT:-KqZevZ25MFzGRa584gU:bkwaek7) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/index.html#L358)):*
> ```HTML
>             <td height="5%">
>               <h3>Block Definition:
>                 <!-- TODO: Separate concerns of format and editable. Add "Edit" button? -->
> ```

Make this a github issue and refer to it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
I don't understand what an Element refers to in this code, which is making it hard to understand the intent of the code.

---

Reviewed 1 of 6 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 20 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 7 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjVlOG2kuORNMxrBM:-KqZjVlOG2kuORNMxrBN:b-g63p4d) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L7)):*
> ```JavaScript
> 
> /**
>  * @fileoverview Block element construction functions.
> ```

This file overview isn't very informative.  "Code for parsing JSON and JavaScript to create Blockly block definitions on a best-effort basis."

or whatever, just needs more information.

---

*[demos/blockfactory/block_definition_extractor.js, line 19 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjnNWLRVIkDOrJrAp:-KqZjnNWLRVIkDOrJrAq:bg35pwv) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L19)):*
>  * @callback nodeChainCallback
?

---

*[demos/blockfactory/block_definition_extractor.js, line 28 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjuOj07Rhy2WRvE3C:-KqZjuOj07Rhy2WRvE3D:bqy9wah) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L28)):*
> ```JavaScript
>  */
> BlockDefinitionExtractor = function() {
>   /** @type {{root: Blockly.Block, current: Blockly.Block}} */
> ```

If possible these should be private.

https://developers.google.com/blockly/guides/modify/web/style-guide#properties

---

*[demos/blockfactory/block_definition_extractor.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZk5b_je81bQMjDFl9:-KqZk5b_je81bQMjDFlA:bsuqf60) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L38)):*
> ```JavaScript
>  * @param {!string} name New element tag name.
>  * @param {Map} name New element's attributes.
>  * @param {string} text New element's inner text.
> ```

@private

---

*[demos/blockfactory/block_definition_extractor.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkTtqZPaxUWYwGUjh:-KqZkTtqZPaxUWYwGUji:bg5m3r7) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L38)):*
> ```JavaScript
>  * @param {!string} name New element tag name.
>  * @param {Map} name New element's attributes.
>  * @param {string} text New element's inner text.
> ```

If text is optional, call it opt_text and annotate correctly.

---

*[demos/blockfactory/block_definition_extractor.js, line 41 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkO_qur3v0DWdX9Ir:-KqZkO_qur3v0DWdX9Is:b-kdfkkg) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L41)):*
> ```JavaScript
>  * @return {Element} The newly created element.
>  */
> BlockDefinitionExtractor.prototype.newElement_ = function(name, attrs, text) {
> ```

`newDomElement_`

---

*[demos/blockfactory/block_definition_extractor.js, line 42 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkA5Qi58jiyaU0XVS:-KqZkA5Qi58jiyaU0XVT:b-oc7stl) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L42)):*
> ```JavaScript
>  */
> BlockDefinitionExtractor.prototype.newElement_ = function(name, attrs, text) {
>   var block1;
> ```

If this won't always be a block, the variable should be named something more generic like `elem`.  If it's always used for blocks, the JSDoc should be updated.

---

*[demos/blockfactory/block_definition_extractor.js, line 43 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkk9fj9jJjMnymmYG:-KqZkk9fj9jJjMnymmYH:bepf6kg) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L43)):*
> ```JavaScript
> BlockDefinitionExtractor.prototype.newElement_ = function(name, attrs, text) {
>   var block1;
>   if(name === 'block') {
> ```

space after if

---

*[demos/blockfactory/block_definition_extractor.js, line 612 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-Kq_FSrsp-Ssg-v7XUaq:-Kq_FSrtunvTvdoysN67:b-ciom25) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L612)):*
> ```JavaScript
>     }
>   }
>   var this_ = this;
> ```

Unused.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235#-:-Kq_FnRR-9-GVpVi_yiT:b-t5yuoi)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 5 of 6 files reviewed at latest revision, 20 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 7 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjVlOG2kuORNMxrBM:-Kr35TvN-ThDZ0H89QAy:bxk3q0x) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L7)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This file overview isn't very informative.  "Code for parsing JSON and JavaScript to create Blockly block definitions on a best-effort basis."

or whatever, just needs more information.
</blockquote></details>

Rewrote.

---

*[demos/blockfactory/block_definition_extractor.js, line 19 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjnNWLRVIkDOrJrAp:-Kr2lnYfZJ2vqv6CMYiI:bss4a2b) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L19)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

>  * @callback nodeChainCallback
?
</blockquote></details>

Removed.

---

*[demos/blockfactory/block_definition_extractor.js, line 28 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjuOj07Rhy2WRvE3C:-Kr2x9qRZ3kn5J84BbMH:bero78j) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L28)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If possible these should be private.

https://developers.google.com/blockly/guides/modify/web/style-guide#properties
</blockquote></details>

Replaced the class with a namespace, leaving only one function public.

---

*[demos/blockfactory/block_definition_extractor.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZk5b_je81bQMjDFl9:-Kr2xHoJAggDyt9pmR7s:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L38)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

@private
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkTtqZPaxUWYwGUjh:-Kr31ykT6iYc8KALDX80:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L38)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If text is optional, call it opt_text and annotate correctly.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 41 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkO_qur3v0DWdX9Ir:-Kr34Gx91iduesXB0kkt:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L41)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

`newDomElement_`
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 42 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkA5Qi58jiyaU0XVS:-Kr34OPc7cnjHkh_NYA2:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L42)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

If this won't always be a block, the variable should be named something more generic like `elem`.  If it's always used for blocks, the JSDoc should be updated.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 43 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkk9fj9jJjMnymmYG:-Kr34QaSt0NaQgbFiNq_:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L43)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

space after if
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 612 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-Kq_FSrsp-Ssg-v7XUaq:-Kr34UvAOAGsBPwct09g:bss4a2b) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L612)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Unused.
</blockquote></details>

Removed.

---

*[demos/blockfactory/factory.js, line 140 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZivDJKaoga4JmmdPN:-Kr34cIMzZ9wk1Y2KUJz:b-4guvl1) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L140)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

So if the `updateBlocksFlag` is set you just don't update the blocks?
</blockquote></details>

I haven't dug deep enough to fully understand the cycles these two global variables  influence, so I can't give you alternative names.

---

*[demos/blockfactory/factory.js, line 143 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZj2Z59J58O3P0ncIr:-Kr34c0BREQWBU9noB6Q:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L143)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

`JSON` should be a string.
</blockquote></details>

Done.

---

*[demos/blockfactory/factory.js, line 192 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZie5x1RMDyZh7Yx5D:-Kr25ybpAMhynhWF9Xdh:blpfzcl) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L192)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Where is `isManual` being used?
</blockquote></details>

Appartently my own cruft. Removed.

---

*[demos/blockfactory/factory.js, line 200 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZiSkPX90S1005c1C0:-Kr2858Q6qlcB2zn5QPZ:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L200)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Make this a helper function that returns a string or throws an error.
</blockquote></details>

Done.

---

*[demos/blockfactory/factory.js, line 232 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfZVoT9-d8ygvtE4f:-Kr287Da8KM8KWf8CpuY:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L232)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

>.<

Discussed offline, and we will not have the Manual JavaScript option until we can spin up an interpreter to do the eval.  The code will remain.
</blockquote></details>

Done.

---

*[demos/blockfactory/factory.js, line 290 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfSJtTL2V482SlgvC:-Kr289m0aS5wsC_mST3N:bfsmx2c) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L290)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Future thought: if it was parseable from the manual format then it's possible to save and link.
</blockquote></details>

Yes, but the linked format will only be the editor workspace XML. That part should be safe (if we fix the various encoding problems).

---

*[demos/blockfactory/factory.js, line 330 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfLHi64AO8c3i_v85:-Kr29N0z3WCl2eY6CEXu:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L330)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

TODO on github and link to issue.
</blockquote></details>

Done.

---

*[demos/blockfactory/factory.js, line 335 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfJ_BCjEHdIQubNI-:-Kr29NXtKPgyvNwuIdni:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L335)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Nit: missing newline at end of file.
</blockquote></details>

Done.

---

*[demos/blockfactory/factory_utils.js, line 88 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfB8DfJF1nRsYB38u:-Kr35OVRElx2O8dOcTzS:b-ryusyk) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory_utils.js#L88)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Add a TODO and github issue to figure out where the infinite loops are coming from and eventually remove this.
</blockquote></details>

Added TODO comment.

---

*[demos/blockfactory/factory_utils.js, line 91 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfG5sBkaIxhBPZ_Ek:-Kr35SC1Oxd0YdSdTVH5:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory_utils.js#L91)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

nit: semicolon after `false`
</blockquote></details>

Done.

---

*[demos/blockfactory/index.html, line 358 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZevZ25MFzGRa584gT:-Kr2kKh0jUGvpiU6rlBU:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/index.html#L358)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Make this a github issue and refer to it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

JC-Orozco commented:
*[demos/blockfactory/factory.js, line 140 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZivDJKaoga4JmmdPN:-Kr6ZTtDVp5IfskSZ40Z:b-vxtfa2) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L140)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I haven't dug deep enough to fully understand the cycles these two global variables  influence, so I can't give you alternative names.
</blockquote></details>

I can shed some light into the reason for updateBlocksFlag and updateBlocksFlagDelay. Beware that Andrew changed the original workflow and reintroduced the Block Definition: manual edit mode. In this explanation I will assume my original version that allowed direct edit on Block Definition: either JSON mode or JavaScript mode. I think we still need this flags to break the round trip conversion loop even with the manual mode reintroduced.
First a conceptual explanation. (Substitute text code for JSON code or JavaScript code depending on the current mode).
The original flow of the Block Factory editor is as follows (triggered by a change in the mainWorkspace):
Block definition in blockly (mainWorkspace) -> text code (JSON or JavaScript) -> block preview (previewWorkspace)
My program adds a converter that takes a preview block and generates a Block definition DOM data structure. So a new flow is introduced as follows (triggered by a change in the code text editor that as per the previous flow generates a preview block):
block preview (previewWorkspace) -> Block definition in blockly (mainWorkspace)
The problem is that since this last flow changes the mainWorkspace the original flow described above gets triggered and an round trip loop was generated that had the effect of deleting the latest code editor change.
My solution to avoid this round loop is introduce this flags updateBlocksFlag and updateBlocksFalgDelayed. I tried first with a single flag but I needed to be sure that two events occurred before clearing the flag. The preview block is generated and the new Block definition in mainWorkspace is generated. So I created two flags, but for some reason I keep getting a complete loop triggered, that is why I introduced a delayed flag which finally solved the round trip problem. I did not have time to learn with much detail the complete flow of the Block Factory application so once I found out that a state flag and a delayed flag solved this problem I moved on. I did think that a delayed flag could be a good idea to avoid too many code to block definition translations which may be hard on the processing power of the computer so a delay may help pace this translation which is generated on each keystroke. But that was just a conceptual idea I did not check if this particular delayed flag improved on that problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

JC-Orozco commented:
In the "Block Definition:" dropdown selector the manual mode is incorrectly labeled as  "Manual JSON...". This could either be Manual JavaScript or Manual JSON depending if the previous mode was JSON or JavaScript

JC-Orozco commented:
*[demos/blockfactory/factory.js, line 140 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZivDJKaoga4JmmdPN:-Kr6rhbCGFYBrPijGa71:b-wwfadb) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L140)):*
<details><summary><i>Previously, JC-Orozco (Juan Carlos Orozco) wrote…</i></summary><blockquote>

I can shed some light into the reason for updateBlocksFlag and updateBlocksFlagDelay. Beware that Andrew changed the original workflow and reintroduced the Block Definition: manual edit mode. In this explanation I will assume my original version that allowed direct edit on Block Definition: either JSON mode or JavaScript mode. I think we still need this flags to break the round trip conversion loop even with the manual mode reintroduced.
First a conceptual explanation. (Substitute text code for JSON code or JavaScript code depending on the current mode).
The original flow of the Block Factory editor is as follows (triggered by a change in the mainWorkspace):
Block definition in blockly (mainWorkspace) -> text code (JSON or JavaScript) -> block preview (previewWorkspace)
My program adds a converter that takes a preview block and generates a Block definition DOM data structure. So a new flow is introduced as follows (triggered by a change in the code text editor that as per the previous flow generates a preview block):
block preview (previewWorkspace) -> Block definition in blockly (mainWorkspace)
The problem is that since this last flow changes the mainWorkspace the original flow described above gets triggered and an round trip loop was generated that had the effect of deleting the latest code editor change.
My solution to avoid this round loop is introduce this flags updateBlocksFlag and updateBlocksFalgDelayed. I tried first with a single flag but I needed to be sure that two events occurred before clearing the flag. The preview block is generated and the new Block definition in mainWorkspace is generated. So I created two flags, but for some reason I keep getting a complete loop triggered, that is why I introduced a delayed flag which finally solved the round trip problem. I did not have time to learn with much detail the complete flow of the Block Factory application so once I found out that a state flag and a delayed flag solved this problem I moved on. I did think that a delayed flag could be a good idea to avoid too many code to block definition translations which may be hard on the processing power of the computer so a delay may help pace this translation which is generated on each keystroke. But that was just a conceptual idea I did not check if this particular delayed flag improved on that problem.

</blockquote></details>

While revisiting this issue, I came up with a theory of why I was not able to break this loop with a single flag. I think the mainWorkspace blocks to code conversion is triggered more than once while doing the preview block to block definition (mainWorkspace DOM) conversion, possibly each time a new node is created on the mainWorkspace DOM or for some other reason, but the posible fact that multiple events are being triggered for a single code to block definition conversion could be the reason the only way I was able to break the loop is with a delayed flag. If this is the case a simple solution would be to use a single state flag again and disable mainWorkspace blocks to code conversion until the complete DOM of the preview to block definition process is generated. Thus avoiding the multiple triggering blocks to code conversion. This could also result in an optimized performance by avoiding unnecessary intermediate blocks to code conversions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
@jc-orozco The drop-down now selects either JSON or JavaScript, not both. There is a separate "Manual JavaScript" mode that has been disabled in anything file local `file://` hosting, because of the security issue surrounding the `eval()` statement.

---

Review status: 2 of 6 files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235#-:-Kr7a_Fk91psirQtIvyh:b-7hzxfh)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 2 of 6 files reviewed at latest revision, 18 unresolved discussions.

---

*[demos/blockfactory/factory.js, line 140 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZivDJKaoga4JmmdPN:-Kr7bPgywiprxjNng4Av:bn21c5q) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory.js#L140)):*
<details><summary><i>Previously, JC-Orozco (Juan Carlos Orozco) wrote…</i></summary><blockquote>

While revisiting this issue, I came up with a theory of why I was not able to break this loop with a single flag. I think the mainWorkspace blocks to code conversion is triggered more than once while doing the preview block to block definition (mainWorkspace DOM) conversion, possibly each time a new node is created on the mainWorkspace DOM or for some other reason, but the posible fact that multiple events are being triggered for a single code to block definition conversion could be the reason the only way I was able to break the loop is with a delayed flag. If this is the case a simple solution would be to use a single state flag again and disable mainWorkspace blocks to code conversion until the complete DOM of the preview to block definition process is generated. Thus avoiding the multiple triggering blocks to code conversion. This could also result in an optimized performance by avoiding unnecessary intermediate blocks to code conversions.
</blockquote></details>

@jc-orozco Thanks for the info.  I think we are going to accept this as is in the sort term.  I will revisit with your notes in mind at some future time.  (As I rush off to copy these notes into the attached issue.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

JC-Orozco commented:
*[demos/blockfactory/block_definition_extractor.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkTtqZPaxUWYwGUjh:-Kr7rwGI7qwt5dAadc_p:b-xiiumo) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L38)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

attrs is replaced by opt_attrs on this definition but not on the second and third use of this parameter in line 93 and 94

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 2 of 6 files reviewed at latest revision, 18 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 38 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkTtqZPaxUWYwGUjh:-Kr8HJsLmg_N02nf_kRi:byag1fa) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L38)):*
<details><summary><i>Previously, JC-Orozco (Juan Carlos Orozco) wrote…</i></summary><blockquote>

attrs is replaced by opt_attrs on this definition but not on the second and third use of this parameter in line 93 and 94
</blockquote></details>

Whoops.  Thanks, JC!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

JC-Orozco commented:
I just read your issue 1268: https://github.com/google/blockly/issues/1268
That will solve this problem and I agree it is an improvement on the UI. You could even clear and gray out the "Manual Edit" checkbox or toggle button on the JavaScript mode and add a small question mark popup message to explain why it is grayed out until the new safe javascript eval is in place. Just an idea.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235#-:-KrBPyq9vUjssH9YinjZ:b7itgrt)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 1 of 1 files at r5.
Review status: 2 of 6 files reviewed at latest revision, 23 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 7 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjVlOG2kuORNMxrBM:-Kr3H_6c-1N_ln1MxMk-:b-mm57db) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L7)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Rewrote.
</blockquote></details>

Move the new fileoverview into here and remove the fileoverview tag + text here.

---

*[demos/blockfactory/block_definition_extractor.js, line 42 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZkA5Qi58jiyaU0XVS:-KrWwbouVVjInPXBEQea:b4igdhv) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L42)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm acking comments that got moved away from their relevant code.  Reviewable didn't handle reordering well this time.

---

*[demos/blockfactory/block_definition_extractor.js, line 52 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWwutgJqluyNHcYfMx:-KrWwutgJqluyNHcYfMy:b-6gadaa) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L52)):*
> ```JavaScript
> BlockDefinitionExtractor.buildBlockFactoryWorkspace = function(block) {
>   var workspaceXml = goog.dom.createDom('xml');
>   var inline = 'AUTO'; // When block.inputsInlineDefault === undefined
> ```

This chunk for figuring out if it's inline or external should be a helper function.  It also doesn't need to be happening here instead of in factoryBase_, where all of the other parsing is happening.

---

*[demos/blockfactory/block_definition_extractor.js, line 58 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWx201T2Wl7SCIQh2H:-KrWx201T2Wl7SCIQh2I:b-w0vtz2) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L58)):*
> ```JavaScript
>     inline = 'EXT';
>   }
>   var connections = 'NONE';
> ```

Same comment for this part, figuring out connection types.

---

*[demos/blockfactory/block_definition_extractor.js, line 87 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWxBRuh1LXP81nt7OS:-KrWxBRuh1LXP81nt7OT:b-sxpnhk) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L87)):*
> ```JavaScript
>  * @param {Map<String,String>} opt_attrs Optional list of attributes.
>  * @param {string?} opt_text Optional inner text.
>  * @return {Element} The newly created element.
> ```

@private 

---

*[demos/blockfactory/block_definition_extractor.js, line 170 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWxvc-h6m4BAloGNMx:-KrWxvc-h6m4BAloGNMy:bg4zw01) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L170)):*
> ```JavaScript
>     var field = fieldRow[i];
>     var fieldDefElement = null;
>     if (field instanceof Blockly.FieldLabel) {
> ```

Make this if-elseif-elseif chain into a separate function that returns the element or null, so that the logic in this parseFields_ function is easier to see.

---

*[demos/blockfactory/block_definition_extractor.js, line 191 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWy8nLXcY7p9j2_wh8:-KrWy8nLXcY7p9j2_wh9:b-iihtx0) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L191)):*
> ```JavaScript
>           field.src_, field.width_, field.height_, field.text_);
>     } else if (field instanceof Blockly.FieldVariable) {
>       // FieldVariable must be before FieldDropdown.
> ```

Is this because FieldVariable extends FieldDropdown?  If so, add that to the comment.

---

*[demos/blockfactory/block_definition_extractor.js, line 197 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyJU_hTjsd48XA5-X:-KrWyJUafxoyLmp0V9Y4:b-1jjjyn) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L197)):*
> ```JavaScript
>       fieldDefElement = BlockDefinitionExtractor.fieldDropdown_(field);
>     }
> ```

What if fieldDefElement is null here?

---

*[demos/blockfactory/block_definition_extractor.js, line 226 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyVDsq5hsx83udZz3:-KrWyVDsq5hsx83udZz4:b2j86rn) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L226)):*
> ```JavaScript
>     var input = block.inputList[i];
>     var align = 'LEFT'; // Left alignment is the default.
>     if (input.align || input.align === 0) {
> ```

Why wrap in this `if`?  It seems like it's just checking the same things as the two internal if statements, and there's no matching `else`.

---

*[demos/blockfactory/block_definition_extractor.js, line 251 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyscMt-ch8RzDsYQx:-KrWyscMt-ch8RzDsYQy:b812psl) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L251)):*
> ```JavaScript
>  *
>  * @param {Blockly.Block} block The example block for the extracted style.
>  * @param {string} connections Define block connections. Options: NONE, LEFT,
> ```

"Define block connections" -> "A string describing which connections the block has.  Options are: ..."

---

*[demos/blockfactory/block_definition_extractor.js, line 288 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWzGs3LnDgqcJLkY83:-KrWzGs3LnDgqcJLkY84:b-xh5v6c) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L288)):*
> ```JavaScript
>   factoryBaseEl.append(helpUrlValue);
> 
>   if (connections === 'LEFT') {
> ```

This should go in the helper that figures out which connections are allowed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 3 of 4 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[demos/blockfactory/factory_utils.js, line 91 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfG5sBkaIxhBPZ_Ek:-KrX-4JtM4EeWBCD6Xb2:b-rgljku) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory_utils.js#L91)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see it yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 7 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZjVlOG2kuORNMxrBM:-KrhIa58neyWB6tTQMoM:b-896fix) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/block_definition_extractor.js#L7)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Move the new fileoverview into here and remove the fileoverview tag + text here.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 52 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWwutgJqluyNHcYfMx:-KrhMBUJXGy4em8eUEmG:b-896fix) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L52)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This chunk for figuring out if it's inline or external should be a helper function.  It also doesn't need to be happening here instead of in factoryBase_, where all of the other parsing is happening.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 58 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWx201T2Wl7SCIQh2H:-KrhNYHT6iR7GPEP8csT:b-896fix) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L58)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Same comment for this part, figuring out connection types.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 87 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWxBRuh1LXP81nt7OS:-KrhNpESWK68nPO68bsG:b-896fix) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L87)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

@private 
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 170 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWxvc-h6m4BAloGNMx:-KrimLOXlBZLAvLFvzR_:b-896fix) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L170)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Make this if-elseif-elseif chain into a separate function that returns the element or null, so that the logic in this parseFields_ function is easier to see.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 191 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWy8nLXcY7p9j2_wh8:-KrhqGu23HdgWTOVzLTY:b-896fix) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L191)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Is this because FieldVariable extends FieldDropdown?  If so, add that to the comment.
</blockquote></details>

Done.

---

*[demos/blockfactory/block_definition_extractor.js, line 197 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyJU_hTjsd48XA5-X:-KrhrDnBaCgvZsqsJU3c:b-ll58w0) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L197)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What if fieldDefElement is null here?
</blockquote></details>

Shouldn't be possible since we've only loaded Blockly defined classes, so I'm now throwing an error in `buildFieldElement_(..)`.

---

*[demos/blockfactory/block_definition_extractor.js, line 226 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyVDsq5hsx83udZz3:-KrhtQy0UnUBkE8NRFS3:b3j607n) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L226)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why wrap in this `if`?  It seems like it's just checking the same things as the two internal if statements, and there's no matching `else`.
</blockquote></details>

@jc-orozco  What was the intention behind this code?

---

*[demos/blockfactory/block_definition_extractor.js, line 251 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyscMt-ch8RzDsYQx:-KrhtapNeCcgojwfWcFL:bvq45yc) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L251)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

"Define block connections" -> "A string describing which connections the block has.  Options are: ..."
</blockquote></details>

Removed `connections` and `inline`.

---

*[demos/blockfactory/block_definition_extractor.js, line 288 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWzGs3LnDgqcJLkY83:-KrimO2eJjF0lrHGHEst:b-896fix) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L288)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This should go in the helper that figures out which connections are allowed.
</blockquote></details>

Done.

---

*[demos/blockfactory/factory_utils.js, line 91 at r3](https://reviewable.io:443/reviews/google/blockly/1235#-KqZfG5sBkaIxhBPZ_Ek:-KrimUJF8wQ6atf6A3Yj:b-trhvdq) ([raw file](https://github.com/google/blockly/blob/07e4bb11e4476f4db690c4849ab20b0f368ac772/demos/blockfactory/factory_utils.js#L91)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I don't see it yet.
</blockquote></details>

Done for realz.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 3 of 6 files reviewed at latest revision, 14 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 271 at r7](https://reviewable.io:443/reviews/google/blockly/1235#-Krkv-JYOHYSudgziX-t:-Krkv-JYOHYSudgziX-u:bx0iarn) ([raw file](https://github.com/google/blockly/blob/2c4f61aa08bbbb0219008f081d385e94a51fbf62/demos/blockfactory/block_definition_extractor.js#L271)):*
> ```JavaScript
>     var input = block.inputList[i];
>     var align = 'LEFT'; // Left alignment is the default.
>     if (input.align || input.align === 0) {
> ```

@rachel-fenichel @jc-orozco This is where that code moved to. (Sorry, I shouldn't have moved it before this was resolved.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 401 at r7](https://reviewable.io:443/reviews/google/blockly/1235#-Krl4hobYhh0f5pxd1QQ:-Krl4hobYhh0f5pxd1QR:b-sgothl) ([raw file](https://github.com/google/blockly/blob/2c4f61aa08bbbb0219008f081d385e94a51fbf62/demos/blockfactory/block_definition_extractor.js#L401)):*
> ```JavaScript
>  * @private
>  */
> BlockDefinitionExtractor.buildFieldLabel = function(text) {
> ```

All of these private functions lost their underscores.  Please put them back.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

JC-Orozco commented:
*[demos/blockfactory/block_definition_extractor.js, line 226 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyVDsq5hsx83udZz3:-KrlRPzx0UazLgMLzjzK:bfcq1lv) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L226)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

@jc-orozco  What was the intention behind this code?
</blockquote></details>

If connection.check_ is undefined then add a Shadow block with null type. There may be more elegant ways to write this, I am avoiding the "can not read length of undefined" error. I do not remember if there are cases when connection.check_ can be undefined because sometimes I write code like this just in case (sort of defensive programming) but maybe I did it because there are cases were connection.check_ can be undefined.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

JC-Orozco commented:
*[demos/blockfactory/block_definition_extractor.js, line 226 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyVDsq5hsx83udZz3:-KrlSyj-pEDRk--Dh_S0:baej2ug) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L226)):*
<details><summary><i>Previously, JC-Orozco (Juan Carlos Orozco) wrote…</i></summary><blockquote>

If connection.check_ is undefined then add a Shadow block with null type. There may be more elegant ways to write this, I am avoiding the "can not read length of undefined" error. I do not remember if there are cases when connection.check_ can be undefined because sometimes I write code like this just in case (sort of defensive programming) but maybe I did it because there are cases were connection.check_ can be undefined.
</blockquote></details>

@rachel-fenichel There is a matching else.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 5 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[demos/blockfactory/block_definition_extractor.js, line 226 at r5](https://reviewable.io:443/reviews/google/blockly/1235#-KrWyVDsq5hsx83udZz3:-KrlfEBSNrA1d4TImbIH:bjilwj6) ([raw file](https://github.com/google/blockly/blob/71bcb407d580252db9dc214847045aba8683b19a/demos/blockfactory/block_definition_extractor.js#L226)):*
<details><summary><i>Previously, JC-Orozco (Juan Carlos Orozco) wrote…</i></summary><blockquote>

@rachel-fenichel There is a matching else.
</blockquote></details>

@jc-orozco the question was about a different if/else, which is now in parseInputs.  Reviewable doesn't handle moving chunks of code around mid-review.  Anyway, it looks like Andrew resolved it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> ship it!

---

Review status: 5 of 6 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235#-:-KrlfWtDnjUp3dxqdxQE:b-quxs7z)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 5 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[demos/blockfactory/block_definition_extractor.js, line 401 at r7](https://reviewable.io:443/reviews/google/blockly/1235#-Krl4hobYhh0f5pxd1QQ:-KrlFkb8RLqFxAojDgIU:b-ebdd44) ([raw file](https://github.com/google/blockly/blob/2c4f61aa08bbbb0219008f081d385e94a51fbf62/demos/blockfactory/block_definition_extractor.js#L401)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

All of these private functions lost their underscores.  Please put them back.
</blockquote></details>

Whoops.  Bad search & replace.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1235)*
<!-- Sent from Reviewable.io -->
